src/v/config/constraints.cc

```diff
@@ -45,6 +52,500 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+  const tristate<T>& tri,
+  const constraint_t& constraint,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property,
+  const std::optional<T> cluster_opt = std::nullopt) {
+    if (tri.is_disabled()) {
+        return false;
+    }
+
+    return validate_value(
+      tri.get_optional(),
+      constraint,
+      topic,
+      topic_property,
+      cluster_property,
+      cluster_opt);
+}
+} // namespace
+
+bool topic_config_satisfies_constraint(
+  const cluster::topic_configuration& topic_cfg,
+  const constraint_t& constraint) {
+    auto partition_count_opt = std::make_optional(topic_cfg.partition_count);
+    auto replication_factor_opt = std::make_optional(
+      topic_cfg.replication_factor);
+    bool ret
+      = validate_value(
+          partition_count_opt,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_partition_count,
+          config::shard_local_cfg().default_topic_partitions.name())
+        || validate_value(
+          replication_factor_opt,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_replication_factor,
+          config::shard_local_cfg().default_topic_replication.name())
+        || validate_value<model::compression>(
+          topic_cfg.properties.compression,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_compression,
+          config::shard_local_cfg().log_compression_type.name(),
+          config::shard_local_cfg().log_compression_type())
+        || validate_value<model::cleanup_policy_bitflags>(
+          topic_cfg.properties.cleanup_policy_bitflags,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_cleanup_policy,
+          config::shard_local_cfg().log_cleanup_policy.name(),
+          config::shard_local_cfg().log_cleanup_policy())
+        || validate_value<model::timestamp_type>(
+          topic_cfg.properties.timestamp_type,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_timestamp_type,
+          config::shard_local_cfg().log_message_timestamp_type.name(),
+          config::shard_local_cfg().log_message_timestamp_type())
+        || validate_value(
+          topic_cfg.properties.segment_size,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_segment_size,
+          config::shard_local_cfg().log_segment_size.name())
+        || validate_value(
+          topic_cfg.properties.retention_bytes,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_retention_bytes,
+          config::shard_local_cfg().retention_bytes.name())
+        || validate_value(
+          topic_cfg.properties.retention_duration,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_retention_duration,
+          config::shard_local_cfg().log_retention_ms.name())
```

Comment:
question: how do you handle a situation for these properties where `-1` is a valid value but represents a positive infinite number?

Scenario: Customer sets a minimum log retention time to be 7 days, but no maximum.  User creates topic with log retention ms to be '-1'.  That would fail validation, correct?

Reply:
The problem was with when to apply -1 for clamp constraints since a clamp is applied when the current value is less than the minimum. But Redpanda assigns -1 (disabled state) at topic creation and -1 is never less than -1. So I went with a design that fails validation here because -1 is not within any range [0, inf+). Of course this will need to be documented.

Reply:
Asked about this in [slack thread](https://redpandadata.slack.com/archives/C04AVDG7N58/p1700167194539089) because depending on semantics, validation could have different results.